library/src/updater.rs

```diff
@@ -411,51 +414,57 @@ pub fn report_launch_failure() -> anyhow::Result<()> {
         };
         // Queue the failure event for later sending since right after this
         // function returns the Flutter engine is likely to abort().
-        state.queue_event(event);
-        // TODO(eseidel): This does the actual save for the above mutations
-        // which is confusing.
-        state
-            .activate_latest_bootable_patch()
-            .map_err(anyhow::Error::from)
+        state.queue_event(event)
     })
 }
 
 pub fn report_launch_success() -> anyhow::Result<()> {
     with_config(|config| {
+        // We can tell the UpdaterState that we have successfully booted from the "next" patch
+        // and make that the "current" patch.
         let mut state =
             UpdaterState::load_or_new_on_error(&config.storage_dir, &config.release_version);
 
-        if let Some(patch) = state.current_boot_patch() {
-            if !state.is_known_good_patch(patch.number) {
-                // Ignore the error here, we'll try to activate the next best patch
-                // even if we fail to mark this one as good.
-                if state.mark_patch_as_good(patch.number).is_ok() {
-                    let config_copy = config.clone();
-                    let client_id = state.client_id_or_default();
-                    std::thread::spawn(move || {
-                        let event = PatchEvent {
-                            app_id: config_copy.app_id.clone(),
-                            arch: current_arch().to_string(),
-                            client_id,
-                            patch_number: patch.number,
-                            platform: current_platform().to_string(),
-                            release_version: config_copy.release_version.clone(),
-                            identifier: EventType::PatchInstallSuccess,
-                        };
-                        let report_result = crate::network::send_patch_event(event, &config_copy);
-                        if let Err(err) = report_result {
-                            error!("Failed to report successful patch install: {:?}", err);
-                        }
-                    });
-                }
-            }
+        let next_boot_patch = match state.next_boot_patch() {
+            Some(patch) => patch,
 
-            state
-                .save()
-                .map_err(|_| anyhow::Error::from(UpdateError::FailedToSaveState))
-        } else {
-            Ok(())
+            // We didn't boot from a patch, so there's nothing to do.
+            None => return Ok(()),
+        };
+
+        let maybe_previous_boot_patch = state.current_boot_patch();
+
+        state.record_boot_success_for_patch(next_boot_patch.number)?;
+
```

Comment:
One could use state.known_good_patch_list() which just returned either [] or [current_boot_patch] and then this code can just check known_good_patch_list().contains(current_boot_patch.number)?

🤷

Reply:
state.is_known_good_patch(number)?